datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -121,58 +291,220 @@ where
+                for (pos, arg) in args.iter().enumerate() {
+                    match arg {
+                        ColumnarValue::Array(arg) => match arg.data_type() {
+                                DataType::Utf8 | DataType::LargeUtf8 => {
+                                    // all good
+                                },
+                                other => return internal_err!("Unsupported data type {other:?} for function {name}, arg # {pos}"),
+                            },
+                        ColumnarValue::Scalar(arg) => { match arg.data_type() {
+                            DataType::Utf8 | DataType::LargeUtf8 => {
+                                // all good
+                            },
+                            other => return internal_err!("Unsupported data type {other:?} for function {name}, arg # {pos}"),
+                        }}
+                    }
+                }
+
+                Ok(ColumnarValue::Array(Arc::new(
+                    strings_to_primitive_function::<i32, O, _, _>(args, op, op2, name)?,
+                )))
+            }
+            other => {
+                internal_err!("Unsupported data type {other:?} for function {name}")
+            }
+        },
+        // if the first argument is a scalar utf8 all arguments are expected to be scalar utf8
+        ColumnarValue::Scalar(scalar) => match scalar {
+            ScalarValue::Utf8(a) | ScalarValue::LargeUtf8(a) => {
+                let mut val: Option<Result<ColumnarValue>> = None;
+                let mut err: Option<DataFusionError> = None;
+
+                match a {
+                    Some(a) => {
+                        // enumerate all the values finding the first one that returns an Ok result
+                        for (pos, v) in args.iter().enumerate().skip(1) {
+                            if let ColumnarValue::Scalar(s) = v {
+                                if let ScalarValue::Utf8(x) | ScalarValue::LargeUtf8(x) =
+                                    s
+                                {
+                                    if let Some(s) = x {
+                                        match op(a.as_str(), s.as_str()) {
+                                            Ok(r) => {
+                                                val = Some(Ok(ColumnarValue::Scalar(
+                                                    S::scalar(Some(op2(r))),
+                                                )));
+                                                break;
+                                            }
+                                            Err(e) => {
+                                                err = Some(e);
+                                            }
+                                        }
+                                    }
+                                } else {
+                                    return internal_err!("Unsupported data type {s:?} for function {name}, arg # {pos}");
```

Comment:
We don't have to support invoking this pattern, but I don't think it should generate an internal error - maybe we can add a test for this too

Reply:
I was thinking about ignoring any argument that are not Utf8/LargeUtf8 as long as there was at least one arg that was acceptable but I didn't think that was a good idea in general as it hides what to me is an obvious error. I'll see if I can get a test to cover this case here.

Reply:
I agree it is not required to support the arguments -- I mostly think the error type should not be "internal error" as that signals a bug in datafusion, which I don't think this is: https://github.com/apache/arrow-datafusion/blob/d14f766b5d3c117d5027f032b38837ec4e285b8a/datafusion/common/src/error.rs#L74-L90

> I was thinking about ignoring any argument that are not Utf8/LargeUtf8 as long as there was at least one arg that was acceptable but I didn't think that was a good idea in general as it hides what to me is an obvious error.

I agree

Reply:
Done, changed to exec_err